find_emails.rb

```diff
@@ -0,0 +1,89 @@
+require 'set'
+require 'nokogiri'
+require 'spidr'
+require 'byebug'
+
+class DomainReader
+  attr_reader :uri, :emails
+
+  def initialize(domain_name)
+    self.uri = domain_name
+  end
+
+  def find_emails(page_limit = nil)
+    @emails = Set.new # use a set to ensure unique emails efficiently
+
+    search = lambda do |page|
+               page.search('//a[@href]').each do |link| # gets all a's with hrefs
+                 link_email = link['href'].match(/\w+@\w+\.\w+/)
+                  # pulls email from href if applicable
+                  # could add further qualifications to regexp in the future
+                  # but this is generally good enough as few hrefs have '@'
+                 emails << link_email if link_email # puts email in set
+               end
+             end
+
+    every_page(search, page_limit)
+  end
+
+  def print_email_addresses
+    if emails.size == 0
+      puts "Found no emails"
+    else
+      puts "\nFound these email addresses:"
+      emails.each { |email| puts email }
+      puts "..."
+    end
+    nil
+  end
+
+  private
+
+  # upon setting the name, validates it and adds the protocol and sub-domain
+  def uri=(name)
```

Comment:
i think it's kind of sneaky that the setting the uri would also validate it and prompt for a new domain. i also wouldn't put that in the constructor.  let's break these two things up so that there's in "input and validation" and then "email finding".  make sense?

theoretically if we wanted to put a web interface on DomainReader, we would have a hard time reusing it the way it is right now, right?